homeassistant/components/tuya/cover.py

```diff
@@ -0,0 +1,140 @@
+from homeassistant.helpers.entity import Entity
+
+from .base import TuyaHaEntity
+from .const import (
+    DOMAIN,
+    TUYA_DEVICE_MANAGER,
+    TUYA_DISCOVERY_NEW,
+    TUYA_HA_DEVICES,
+    TUYA_HA_TUYA_MAP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+TUYA_SUPPORT_TYPE = {"cl", "clkg"}  # Curtain  # Curtain Switch
+
+# Curtain
+# https://developer.tuya.com/en/docs/iot/f?id=K9gf46o5mtfyc
+DPCODE_CONTROL = "control"
+DPCODE_PERCENT_CONTROL = "percent_control"
+DPCODE_PERCENT_STATE = "percent_state"
+DPCODE_SITUATION_SET = "situation_set"
+
+ATTR_POSITION = "position"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up tuya cover dynamically through tuya discovery."""
+    hass.data[DOMAIN][entry.entry_id][TUYA_HA_TUYA_MAP][
+        DEVICE_DOMAIN
+    ] = TUYA_SUPPORT_TYPE
+
+    @callback
+    def async_discover_device(dev_ids: list[str]) -> None:
+        """Discover and add a discovered tuya cover."""
+        _LOGGER.debug("cover add-> %s", dev_ids)
+        if not dev_ids:
+            return
+
+        entities = _setup_entities(hass, entry, dev_ids)
+
+        async_add_entities(entities)
+
+    entry.async_on_unload(
+        async_dispatcher_connect(
+            hass, TUYA_DISCOVERY_NEW.format(DEVICE_DOMAIN), async_discover_device
+        )
+    )
+
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    device_ids = []
+    for (device_id, device) in device_manager.device_map.items():
+        if device.category in TUYA_SUPPORT_TYPE:
+            device_ids.append(device_id)
+    async_discover_device(device_ids)
+
+
+def _setup_entities(
+    hass: HomeAssistant, entry: ConfigEntry, device_ids: list[str]
+) -> list[Entity]:
+    """Set up Tuya Cover."""
+    device_manager = hass.data[DOMAIN][entry.entry_id][TUYA_DEVICE_MANAGER]
+    entities: list[Entity] = []
+    for device_id in device_ids:
+        device = device_manager.device_map[device_id]
+        if device is None:
+            continue
+        entities.append(TuyaHaCover(device, device_manager))
+        hass.data[DOMAIN][entry.entry_id][TUYA_HA_DEVICES].add(device_id)
+    return entities
+
+
+class TuyaHaCover(TuyaHaEntity, CoverEntity):
+    """Tuya Switch Device."""
+
+    _attr_device_class = DEVICE_CLASS_CURTAIN
+
+    @property
+    def is_closed(self) -> bool | None:
+        """Return is cover is closed."""
+        return None
+
+    @property
+    def current_cover_position(self) -> int:
+        """Return cover current position."""
+        position = self.tuya_device.status.get(DPCODE_PERCENT_STATE, 0)
+        if DPCODE_SITUATION_SET not in self.tuya_device.status:
+            return 1 + int(0.98 * (100 - position))
+        if self.tuya_device.status.get(DPCODE_SITUATION_SET) == "fully_open":
+            return 1 + 0.98 * position
+
+        return 1 + int(0.98 * (100 - position))
```

Comment:
0 means closed in Home Assistant. With this formula you will never be able to return 0.

Reply:
Because in the tuya cover, some cover status 0 represents fully open, which makes it impossible to close in the home assistant. At present, there is no good way to judge which cover status 0 represents open in the home assistant, so we use the range of 1- 99, so that even if it reaches 1, it can continue to operate

Reply:
@balloob as we just discuss with Martin in the Zoom meeting, Tuya's cover devices have different rules definition for movement controlling. Some curtains use 0 status to represent open, some use 100 to represent it. We cannot provide a perfect cover driver feature to satisfy all of PBT products, this is a challenge for us. Here are the suggested option:
1. We release the cover platform now and improve it based on users' feedback to support more PBT cover devices. Make it better with the help of community.
2. If the cover platform's implementation doesn't follow Home Assistant's rule and it's not suggested to provide to our users, we can hold it until we find a better solution. But this will take much more time and we don't have a better solution for now.